### PR TITLE
[gh-6980] Client: clean up old allocs before running new ones  using the `exec` task driver.

### PR DESCRIPTION
**Pull request by Juanadelacuesta**
Whenever the "exec" task driver is being used, nomad runs a plug in that in time runs the task on a container under the hood. If by any circumstance the executor is killed, the task is reparented to the init service and wont be stopped by Nomad in case of a job updated or stop.

This PR introduces two mechanisms to avoid this behaviour: 

- Adds signal catching and handling to the executor, so in case of a SIGTERM, the signal will also be passed on to the task.
- Adds a pre start clean up of the processes in the container, ensuring only the ones the executor runs are present at any given time.

It fixes [#6980](https://github.com/hashicorp/nomad/issues/6980)

### CODE REVIEW

File: drivers/shared/executor/executor_linux.go [hunk 1]
```diff
@@ -188,7 +213,20 @@ func (l *LibcontainerExecutor) Launch(command *ExecCommand) (*ProcessState, erro
...
 
-	go l.wait()
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		defer cancel()
+
+		go l.wait()
+
+		signal := l.catchSignals(ctx)
```
Comment (tgross) [inline drivers/shared/executor/executor_linux.go:223]: Once we call `go l.wait()` we're continuing execution to the signal handler, but this thread could at this point jump to the `go l.wait()` goroutine and have it return before our signal handler is set up. I think we want to set up the signal handler first and then wait?

File: go.mod [hunk 2]
```diff
@@ -172,7 +172,7 @@ require (
 	github.com/cespare/xxhash/v2 v2.2.0 // indirect
 	github.com/checkpoint-restore/go-criu/v5 v5.3.0 // indirect
 	github.com/cheggaaa/pb/v3 v3.0.5 // indirect
-	github.com/cilium/ebpf v0.9.1 // indirect
+	github.com/cilium/ebpf v0.11.0 // indirect
```
[the remark this answers is not in the document]
Comment (Juanadelacuesta) [reply go.mod:175]: I tested a few things before, this is probably a left over of one of those attempts.

File: drivers/shared/executor/executor_linux.go [hunk 3]
```diff
@@ -97,6 +98,27 @@ func (l *LibcontainerExecutor) ListProcesses() *set.Set[int] {
...
 
+// cleanOldProcessesInCGroup kills processes that might ended up orphans when the
+// executor was unexpectedly killed and nomad can't reconnect to them.
+func (l *LibcontainerExecutor) cleanOldProcessesInCGroup(nomadRelativePath string) {
+	l.logger.Info("removing old processes", "path", nomadRelativePath)
+
+	root := cgroupslib.GetDefautlRoot()
+	orphansPIDs, err := cgroups.GetAllPids(filepath.Join(root, nomadRelativePath))
+	if err != nil {
+		l.logger.Error("unable to get orphan allocs PIDs", err)
```
[the remark this answers is not in the document]
Comment (Juanadelacuesta) [reply drivers/shared/executor/executor_linux.go:109]: If there are no processes, wi will get an empty slice, but no error, so we wont print the error message.

File: drivers/shared/executor/executor_linux.go [hunk 4]
```diff
@@ -97,6 +124,34 @@ func (l *LibcontainerExecutor) ListProcesses() *set.Set[int] {
...
+	orphansPIDs, err := cgroups.GetAllPids(filepath.Join(root, nomadRelativePath))
+	if err != nil {
+		l.logger.Error("unable to get orphaned task PIDs", "error" err)
+		return
+	}
+
+	for _, pid := range orphansPIDs {
+		l.logger.Info("killing orphaned process", "pid", pid)
+
+		// Avoid bringind down the whole node by mistake, very unlikly case,
```
Comment (tgross) [inline drivers/shared/executor/executor_linux.go:142]:
```suggestion
		// Avoid bringing down the whole node by mistake, very unlikely case,
```

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgross) [reply go.mod:175]: Ok, we'll want to back this change out before merging.

File: drivers/shared/executor/executor_linux.go [hunk 5]
```diff
@@ -78,17 +79,43 @@ type LibcontainerExecutor struct {
...
+
+func (l *LibcontainerExecutor) catchSignals() {
+	l.logger.Trace("waiting for signals")
+	defer signal.Stop(l.sigChan)
+	defer close(l.sigChan)
+
+	signal.Notify(l.sigChan, syscall.SIGHUP, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGINT, syscall.SIGSEGV)
+	for {
+		signal := <-l.sigChan
+		if signal == syscall.SIGTERM || signal == syscall.SIGKILL || signal == syscall.SIGINT {
```
Comment (tgross) [inline drivers/shared/executor/executor_linux.go:93]: `SIGKILL` isn't one of the signals we've masked above, and we can't catch it anyways. So this block really is only translating `SIGTERM` to `SIGINT`, right?

**Review by tgross — COMMENTED** [review]
Overall this is looking great @Juanadelacuesta. I've left one question about the signals we're masking, and then there's a couple remaining bits to mop up like the `go.mod` and then I think this will be ready to ship.

**Review by tgross — APPROVED** [review]
LGTM! Nice work on this one!

**Pull request merged**